src/System.Management.Automation/utils/unicode/UnicodeSimpleCaseFolding-g2.cs

```diff
@@ -0,0 +1,807 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+
+namespace System.Management.Automation.Unicode
+{
+    /// <summary>
+    /// Unicode Simple Case Folding.
+    /// </summary>
+    internal static partial class SimpleCaseFolding
+    {
+        //private static ref ushort refL1 => ref L1[0];
+        //private static ref ushort refL3 => ref L3[0];
+        //private static ushort refL1 = L3[0];
+        //private static ushort refL3 = L3[0];
+
+        /// <summary>
+        /// Simple case folding of the char (Utf16).
+        /// </summary>
+        /// <param name="c">Source char.</param>
+        /// <returns>
+        /// Returns folded char.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char SimpleCaseFold(char c)
+        {
+            //var v = Unsafe.Add(ref refL1, 1);
+            //var ch = Unsafe.Add(ref refL3, v + (c & 0xFF));
+            var v = Unsafe.Add(ref L1[0], c >> 8);
+            var ch = Unsafe.Add(ref L3[0], v + (c & 0xFF));
+
+            //var v = L1[c >> 8];
+            //var ch = L3[v + (c & 0xFF)];
+            //ref ushort L1a = ref L1[0];
+            //ref ushort L3a = ref L3[0];
+            //var v = Unsafe.Add(ref L1a, c >> 8);
+            //var ch = Unsafe.Add(ref L3a, v + (c & 0xFF));
+            //var v = Unsafe.Add(ref refL1, c >> 8);
+            //var ch = Unsafe.Add(ref refL3, v + (c & 0xFF));
+            //var ch = Unsafe.Add(ref refL3, Unsafe.Add(ref refL1, c >> 8) + (c & 0xFF));
+
+            return ch == 0 ? c : Unsafe.As<ushort, char>(ref ch);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char SimpleCaseFold1(char c)
+        {
+            ushort v    = L1[c >> 8];
+            v           = L3[v + (c & 0xFF)];
+
+            return v == 0 ? c : (char)v;
+        }
+
+        /// <summary>
+        ///  Simple case folding of the string.
+        /// </summary>
+        /// <param name="source">Source string.</param>
+        /// <returns>
+        /// Returns folded string.
+        /// </returns>
+        //[MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
There is a lot of commented code and attributes in this file.  Did you write this all yourself?

Reply:
I started with zero experience in Unicode internals and c# in-depth optimizations. For the last two months I have been moving in small steps so that the code contains a lot of commented code which reflects my attempts to find fastest code. I'll remove its after the code will become stable.
All this code is written by me except for gen.cs that come from @tarekgh's gist.